druid-shell/src/platform/mac/keyboard.rs

```diff
@@ -0,0 +1,381 @@
+        Code::PageUp => Key::PageUp,
+        Code::PageDown => Key::PageDown,
+        Code::ArrowLeft => Key::ArrowLeft,
+        Code::ArrowRight => Key::ArrowRight,
+        Code::ArrowUp => Key::ArrowUp,
+        Code::ArrowDown => Key::ArrowDown,
+        Code::Enter => Key::Enter,
+        Code::NumpadEnter => Key::Enter,
+        Code::Help => Key::Help,
+        _ => return None,
+    })
+}
+
+fn is_valid_key(s: &str) -> bool {
+    match s.chars().next() {
+        None => false,
+        Some(c) => c >= ' ' && c != '\x7f' && !('\u{e000}'..'\u{f900}').contains(&c),
+    }
+}
+
+fn is_modifier_code(code: Code) -> bool {
+    match code {
+        Code::ShiftLeft
+        | Code::ShiftRight
+        | Code::AltLeft
+        | Code::AltRight
+        | Code::ControlLeft
+        | Code::ControlRight
+        | Code::MetaLeft
+        | Code::MetaRight
+        | Code::CapsLock
+        | Code::Help => true,
+        _ => false,
+    }
+}
+
+impl KeyboardState {
+    pub fn new() -> KeyboardState {
+        let last_mods = NSEventModifierFlags::empty();
+        KeyboardState { last_mods }
+    }
+
+    pub fn process_native_event(&mut self, event: id) -> Option<KeyboardEvent> {
+        unsafe {
+            let event_type = event.eventType();
+            let key_code = event.keyCode();
+            let code = key_code_to_code(key_code);
+            let location = code_to_location(code);
+            let raw_mods = event.modifierFlags();
+            let modifiers = make_modifiers(raw_mods);
+            let state = match event_type {
+                NSEventType::NSKeyDown => KeyState::Down,
+                NSEventType::NSKeyUp => KeyState::Up,
+                NSEventType::NSFlagsChanged => {
+                    // We use `bits` here because we want to distinguish the
+                    // device dependent bits (when both left and right keys
+                    // may be pressed, for example).
+                    let any_down = raw_mods.bits() & !self.last_mods.bits();
+                    self.last_mods = raw_mods;
+                    if is_modifier_code(code) {
+                        if any_down == 0 {
+                            KeyState::Up
+                        } else {
+                            KeyState::Down
+                        }
+                    } else {
+                        // HandleFlagsChanged has some logic for this; it might
+                        // happen when an app is deactivated by Command-Tab. In
+                        // that case, the best thing to do is synthesize the event
+                        // from the modifiers. But a challenge there is that we
+                        // might get multiple events.
+                        return None;
+                    }
+                }
+                _ => unreachable!(),
+            };
+            let is_composing = false;
+            let repeat: bool = event_type == NSEventType::NSKeyDown && msg_send!(event, isARepeat);
```

Comment:
total nit but convention is braces for `msg_send!`, which I like because it looks like objc.
```suggestion
            let repeat: bool = event_type == NSEventType::NSKeyDown && msg_send![event, isARepeat];
```